staging/vhost-device-gpu/src/device.rs

```diff
@@ -710,351 +708,461 @@ impl VhostUserBackend for VhostUserGpuBackend {
+    use crate::virtio_gpu::MockVirtioGpu;
+    use mockall::predicate;
+
+    use crate::protocol::{
+        VIRTIO_GPU_CMD_RESOURCE_CREATE_2D, VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM,
+        VIRTIO_GPU_RESP_ERR_UNSPEC, VIRTIO_GPU_RESP_OK_NODATA,
+    };
+    use assert_matches::assert_matches;
     use std::{
-        collections::BTreeMap,
-        mem::size_of,
-        sync::{Arc, Mutex},
+        fs::File,
+        io::ErrorKind,
+        iter::zip,
+        mem,
+        os::{fd::FromRawFd, unix::net::UnixStream},
+        sync::Arc,
     };
     use vhost_user_backend::{VringRwLock, VringT};
-    use virtio_bindings::bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
-    use virtio_queue::{mock::MockSplitQueue, Descriptor, DescriptorChain, Queue};
+    use virtio_bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
+    use virtio_queue::{mock::MockSplitQueue, Descriptor, Queue, QueueT};
     use vm_memory::{
-        Address, ByteValued, Bytes, GuestAddress, GuestMemoryAtomic, GuestMemoryLoadGuard,
-        GuestMemoryMmap,
+        ByteValued, Bytes, GuestAddress, GuestMemory, GuestMemoryAtomic, GuestMemoryMmap,
     };
 
-    type GpuDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
-
     const SOCKET_PATH: &str = "vgpu.socket";
+    const MEM_SIZE: usize = 32 * 1024 * 1024; // 32MiB
 
-    #[derive(Copy, Clone, Default)]
-    #[repr(C)]
-    struct VirtioGpuOutHdr {
-        a: u16,
-        b: u16,
-        c: u32,
-    }
-
-    // SAFETY: The layout of the structure is fixed and can be initialized by
-    // reading its content from byte array.
-    unsafe impl ByteValued for VirtioGpuOutHdr {}
-
-    #[derive(Copy, Clone, Default)]
-    #[repr(C)]
-    struct VirtioGpuInHdr {
-        d: u8,
-    }
-
-    // SAFETY: The layout of the structure is fixed and can be initialized by
-    // reading its content from byte array.
-    unsafe impl ByteValued for VirtioGpuInHdr {}
+    const CURSOR_QUEUE_ADDR: GuestAddress = GuestAddress(0x0);
+    const CURSOR_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x1_000);
+    const CURSOR_QUEUE_SIZE: u16 = 16;
+    const CONTROL_QUEUE_ADDR: GuestAddress = GuestAddress(0x2_000);
+    const CONTROL_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x10_000);
+    const CONTROL_QUEUE_SIZE: u16 = 1024;
 
-    fn init() -> (
-        Arc<VhostUserGpuBackend>,
-        GuestMemoryAtomic<GuestMemoryMmap>,
-        VringRwLock,
-    ) {
+    fn init() -> (Arc<VhostUserGpuBackend>, GuestMemoryAtomic<GuestMemoryMmap>) {
         let backend = VhostUserGpuBackend::new(GpuConfig::new(
             SOCKET_PATH.into(),
             GpuMode::ModeVirglRenderer,
         ))
         .unwrap();
         let mem = GuestMemoryAtomic::new(
-            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x1000)]).unwrap(),
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), MEM_SIZE)]).unwrap(),
```

Comment:
was there any specific reason for increasing the memory size? if none maybe it's too much?

Reply:
Hmm, but it seems to work fine 2Mib instead of 32Mib, so I'll decrease it.